internal/base/base.go

```diff
@@ -655,9 +655,101 @@ func DecodeSchedulerEnqueueEvent(b []byte) (*SchedulerEnqueueEvent, error) {
 	}, nil
 }
 
+// Deadlines is a collection that holds deadlines for all active tasks.
+// When a deadline is reached, the associated cancel function is executed.
+// Assumes that deadlines have unique IDs and will not be added multiples times.
+//
+// Deadlines is safe for concurrent use by multiple goroutines.
+type Deadlines struct {
+	update chan deadline
+}
+
+// NewDeadlines returns a Deadlines instance.
+func NewDeadlines(abort chan struct{}, size int) *Deadlines {
+	d := &Deadlines{update: make(chan deadline, size)}
+	go func() {
+		dls := []deadline{}
+		for {
+			if func() bool {
+				var c <-chan time.Time
+				if len(dls) == 0 {
+					c = make(chan time.Time) // dummy
+				} else {
+					t := time.NewTimer(time.Until(dls[0].dl))
+					defer t.Stop()
+					c = t.C
+				}
+				select {
```

Comment:
The timer could be created once, then reset appropriately:

```suggestion
	go func() {
		dls := []deadline{}
		t := time.NewTimer(time.Hour) // initialize timer with dummy duration
		t.Stop()

		for {
			if func() bool {
				if len(dls) > 0 {
					t.Reset(time.Until(dls[0].dl))
					defer t.Stop()
				}
				c := t.C
				select {
```